[Security Solution] `FinalEdit`: Add fields that are common for all rule types

nikitaindik opened the pull request:
**Partially addresses: https://github.com/elastic/kibana/issues/171520**
**Is a follow-up to: https://github.com/elastic/kibana/pull/196326**

This PR enables editing of common fields in the new "Updates" tab of the rule upgrade flyout. The common fields are fields applicable to all rule types.

## Summary
These fields are editable now:
 - `building_block`
 - `description`
 - `false_positives`
 - `investigation_fields`
 - `max_signals`
 - `note`
 - `references`
 - `related_integrations`
 - `required_fields`
 - `risk_score`
 - `risk_score_mapping`
 - `rule_name_override`
 - `rule_schedule`
 - `setup`
 - `severity`
 - `severity_mapping`
 - `tags`
 - `threat`
 - `timeline_template`
 - `timestamp_override`

<img width="2672" alt="Scherm­afbeelding 2024-10-16 om 17 32 06" src="https://github.com/user-attachments/assets/6dd615e2-6e84-4e1f-b674-f42d03f575e7">

### Testing
 - Ensure the `prebuiltRulesCustomizationEnabled` feature flag is enabled.
 - To simulate the availability of prebuilt rule upgrades, downgrade a currently installed prebuilt rule using the `PATCH api/detection_engine/rules` API. 
   - Set `version: 1` in the request body to downgrade it to version 1.
   - Modify other rule fields in the request body as needed to test the changes.

<!--ONMERGE {"backportTargets":["8.x"]} ONMERGE-->

nikitaindik committed:
Add `tags`

nikitaindik committed:
Add `description`

nikitaindik committed:
Add `references`

nikitaindik committed:
Add `false_positives`

nikitaindik committed:
Add `investigation_fields`

elasticmachine commented:
<!-- buildkite-pr-bot-trigger-comment -->:robot: Jobs for this PR can be triggered through checkboxes. :construction: 

:information_source: To trigger the CI, please tick the checkbox below :point_down: 

- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-pull-request --> Click to trigger **kibana-pull-request** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-project-from-pr --> Click to trigger **kibana-deploy-project-from-pr** for this PR!

nikitaindik committed:
Add `building_block`

nikitaindik committed:
Add `max_signals`

nikitaindik committed:
Add `note`

nikitaindik committed:
Add `setup`

nikitaindik committed:
Add `related_integrations`

nikitaindik committed:
Add `rule_name_override`

nikitaindik committed:
Add `rule_schedule`

nikitaindik committed:
Add `threat`

nikitaindik committed:
Add `timeline_template`

nikitaindik committed:
Add `timestamp_override`

nikitaindik committed:
Add `required_fields`

nikitaindik committed:
Add `risk_score` and `risk_score_mapping`

nikitaindik committed:
Add `severity`, `severity_override`, `risk_score` and `risk_score_override`

nikitaindik committed:
Merge remote-tracking branch 'upstream/main' into editable-common-fields-2

nikitaindik committed:
Update import paths

nikitaindik committed:
Update serializer and deserializer to match field names

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/247775)
* Commit: fcd9457b531dd1214b11935474c21ebfa167031e

### Failed CI Steps
* [FTR Configs #18](https://buildkite.com/elastic/kibana-pull-request/builds/247775#0192e836-5730-4ca2-bbb0-6dfcfda27453)
* [FTR Configs #18](https://buildkite.com/elastic/kibana-pull-request/builds/247775#0192e85d-d8e5-4ce1-9c9c-20f814921a90)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/247775#0192e836-5730-4ca2-bbb0-6dfcfda27453) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/247775/jobs/0192e836-5730-4ca2-bbb0-6dfcfda27453/artifacts/0192e85d-941b-400d-bef4-24b7e972e9e9) FTR Configs #<span></span>18 / Cloud Security Posture Security Alerts Page - Graph visualization should render graph visualization
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/247775#0192e85d-d8e5-4ce1-9c9c-20f814921a90) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/247775/jobs/0192e85d-d8e5-4ce1-9c9c-20f814921a90/artifacts/0192e874-a211-4941-9898-154b6af1e229) FTR Configs #<span></span>18 / Cloud Security Posture Security Alerts Page - Graph visualization should render graph visualization

cc @nikitaindik

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/247775","number":247775,"commit":"fcd9457b531dd1214b11935474c21ebfa167031e"}],"number":247775}
buildkite-pr-comment-->

maximpn reviewed (commented):
@nikitaindik I had a look at the diff. It's huge effort 💪 

I noticed non critical issues and left comments about them. It could be summarized into the following

- some functions in particular related to mappings are unclear
- there are potential performance issues due to `aria` attributes instance changing every render
- some properties don't look necessary
- it's good to have function's return value type

I'm gonna test each field addressed in this PR and post a separate comment about that.

nikitaindik committed:
Merge remote-tracking branch 'upstream/main' into editable-common-fields-2

nikitaindik committed:
Refactor `getFlattenedArrayFieldNames`

nikitaindik commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/components/risk_score_mapping/default_risk_score.tsx:
```diff
@@ -0,0 +1,86 @@
+  onChange,
+  errorMessage,
+  idAria,
+  dataTestSubj = 'defaultRiskScore',
```
Upon investigation I decided to skip this one since I'll then have to make the prop values more generic and update a bunch of places on the Rule Editing page and in tests for consistency. That's going to be a lot of changes.

maximpn commented:
I've tested the PR locally and the majority of the editable fields work as expected. Though I noticed the following problems

- Max Signals allows to save any value despite a validation error show

https://github.com/user-attachments/assets/c42983d7-fc8f-43f6-9663-7b4d413537e4

- Rule name override doesn't show any values (there is no data in rule indices locally so Kibana isn't able to fetch available values)

https://github.com/user-attachments/assets/fb3b895a-bb50-4020-b58a-f1bbd8f66edb

We should show the current value and provide an ability to remove it so final value doesn't have rule name override.

- Timestamp override doesn't show any values (there is no data in rule indices locally so Kibana isn't able to fetch available values)

https://github.com/user-attachments/assets/9de9ac2c-aed8-41e1-9e3b-757f8fce804e

We should show the current value and provide an ability to remove it so final value doesn't have rule name override.


Additionally I've noticed it's quite hard to understand what data source the prebult rule uses to ingest data in proper indices. Though it's rather a nice to have feature and out of scope of this PR.

nikitaindik committed:
Address PR feedback

nikitaindik committed:
Fix `max_signals` validation

nikitaindik committed:
Merge remote-tracking branch 'upstream/main' into editable-common-fields-2

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/249511)
* Commit: db15ed8312834de4ef8c504e6cbef7288d859bb6

### Failed CI Steps
* [Quick Checks](https://buildkite.com/elastic/kibana-pull-request/builds/249511#01930b6b-ccc5-47bd-8840-d2f8c91ac932)

### History
* :broken_heart: [Build #247775](https://buildkite.com/elastic/kibana-pull-request/builds/247775) failed fcd9457b531dd1214b11935474c21ebfa167031e

cc @nikitaindik

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/249511","number":249511,"commit":"db15ed8312834de4ef8c504e6cbef7288d859bb6"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/247775","number":247775,"commit":"fcd9457b531dd1214b11935474c21ebfa167031e"}],"number":249511}
buildkite-pr-comment-->

nikitaindik committed:
Update i18n

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/249536)
* Commit: 6986e21bd8362b1b15dec65018798ff39f29e7c6

### Failed CI Steps
* [Defend Workflows Cypress Tests #9](https://buildkite.com/elastic/kibana-pull-request/builds/249536#01930bb6-33ca-4164-85e2-540448adc3a4)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/249536#01930bb6-33ca-4164-85e2-540448adc3a4) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/249536/jobs/01930bb6-33ca-4164-85e2-540448adc3a4/artifacts/01930bec-b014-4e98-9712-b6e12ece353b) Defend Workflows Cypress Tests #<span></span>9 / No License User cannot use endpoint action in form response actions are disabled response actions are disabled

### History
* :broken_heart: [Build #249511](https://buildkite.com/elastic/kibana-pull-request/builds/249511) failed db15ed8312834de4ef8c504e6cbef7288d859bb6
* :broken_heart: [Build #247775](https://buildkite.com/elastic/kibana-pull-request/builds/247775) failed fcd9457b531dd1214b11935474c21ebfa167031e

cc @nikitaindik

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/249536","number":249536,"commit":"6986e21bd8362b1b15dec65018798ff39f29e7c6"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/249511","number":249511,"commit":"db15ed8312834de4ef8c504e6cbef7288d859bb6"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/247775","number":247775,"commit":"fcd9457b531dd1214b11935474c21ebfa167031e"}],"number":249536}
buildkite-pr-comment-->

nikitaindik committed:
Refactor risk score field error handling and fix `maxSignals` validation on Rule creation page

nikitaindik committed:
Fix a test for `default_risk_score`

nikitaindik committed:
Merge remote-tracking branch 'upstream/main' into editable-common-fields-2

nikitaindik commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation/components/required_fields/utils.ts:
```diff
@@ -26,3 +28,13 @@ export function pickTypeForName({ name, type, typesByFieldName = {} }: PickTypeF
   return typesAvailableForName[0] ?? type;
 }
+
+export function getFlattenedArrayFieldNames(
```
Thanks! I've added an explanatory comment to the function and to the location where it's called.

nikitaindik commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/components/risk_score_mapping/default_risk_score.tsx:
```diff
@@ -0,0 +1,86 @@
+}: DefaultRiskScoreProps) {
+  const handleChange = useCallback<NonNullable<EuiRangeProps['onChange']>>(
+    (event) => {
+      const newValue = (event.target as HTMLInputElement).value;
```
It turns out I can't use `event.currentTarget` because it causes our tests to fail, even though everything still works in the UI. Therefore, I had to revert to the `event.target` approach.

nikitaindik commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/components/risk_score_mapping/default_risk_score.tsx:
```diff
@@ -0,0 +1,86 @@
+  return typeof value === 'number' && value >= MIN_RISK_SCORE && value <= MAX_RISK_SCORE;
+}
+
+export function validateDefaultRiskScore(value: unknown) {
```
Refactored and moved the reusable part to `x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/validators`

nikitaindik commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/components/risk_score_mapping/risk_score_override.tsx:
```diff
@@ -0,0 +1,163 @@
+        >
+          <EuiFlexItem grow={false}>
+            <EuiCheckbox
+              id={`risk_score-mapping-override`}
```
Thanks. Fixed.

nikitaindik commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/components/risk_score_mapping/risk_score_override.tsx:
```diff
@@ -0,0 +1,163 @@
+    );
+  }, [isDisabled, isMappingChecked, onToggleMappingChecked]);
+
+  const fieldTypeFilter = useMemo(() => ['number'], []);
```
Thanks. Fixed.

nikitaindik commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/components/risk_score_mapping/risk_score_override.tsx:
```diff
@@ -0,0 +1,163 @@
+      }
+      fullWidth
+      data-test-subj={`${dataTestSubj}-riskOverride`}
+      describedByIds={idAria ? [idAria] : undefined}
```
Thanks. Fixed.

nikitaindik commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/components/risk_score_mapping/risk_score_override.tsx:
```diff
@@ -0,0 +1,163 @@
+      <NestedContent>
+        <EuiSpacer size="s" />
+        {isMappingChecked && (
+          <EuiFlexGroup direction={'column'} gutterSize="s">
```
Thanks. Fixed.

nikitaindik commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/components/risk_score_mapping/risk_score_override.tsx:
```diff
@@ -0,0 +1,163 @@
+              <EuiFlexGroup alignItems="center" gutterSize="s">
+                <EuiFlexItemComboBoxColumn>
+                  <FieldComponent
+                    placeholder={''}
```
Thanks. Fixed.

nikitaindik commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/components/risk_score_mapping/risk_score_override.tsx:
```diff
@@ -0,0 +1,163 @@
+                    indexPattern={indices}
+                    selectedField={selectedField}
+                    fieldTypeFilter={fieldTypeFilter}
+                    isLoading={false}
```
Thanks. Fixed.

nikitaindik commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/components/risk_score_mapping/risk_score_override.tsx:
```diff
@@ -0,0 +1,163 @@
+                  />
+                </EuiFlexItemComboBoxColumn>
+                <EuiFlexItemIconColumn grow={false}>
+                  <EuiIcon type={'sortRight'} />
```
Thanks. Fixed.

nikitaindik commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/components/risk_score_mapping/risk_score_override.tsx:
```diff
@@ -0,0 +1,163 @@
+                  <EuiIcon type={'sortRight'} />
+                </EuiFlexItemIconColumn>
+                <EuiFlexItemRiskScoreColumn grow={false}>
+                  <EuiText size={'s'}>{i18n.RISK_SCORE_FIELD}</EuiText>
```
Thanks. Fixed.

nikitaindik commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/components/risk_score_mapping/translations.tsx:
```diff
@@ -63,3 +64,11 @@ export const RISK_SCORE_MAPPING_DETAILS = i18n.translate(
 );
+
+export const RISK_SCORE_VALIDATION_ERROR = i18n.translate(
+  'xpack.securitySolution.alerts.defaultRiskScore.validationError',
```
Good idea. I've added it to the refactoring list for later.

nikitaindik commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/components/severity_mapping/default_severity.tsx:
```diff
@@ -0,0 +1,58 @@
+        label={<DefaultSeverityLabel />}
+        fullWidth
+        data-test-subj="detectionEngineStepAboutRuleSeverity"
+        describedByIds={['detectionEngineStepAboutRuleSeverity']}
```
Thanks. Fixed.

nikitaindik commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/components/severity_mapping/severity_override.tsx:
```diff
@@ -0,0 +1,201 @@
+} from '@elastic/eui';
+import { noop } from 'lodash/fp';
+import React, { useMemo } from 'react';
+import styled from 'styled-components';
```
Thanks. I refactored it to use @emotion/css.

nikitaindik commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/components/severity_mapping/severity_override.tsx:
```diff
@@ -0,0 +1,201 @@
+  onFieldChange: (index: number, severity: Severity, [newField]: DataViewFieldBase[]) => void;
+  onFieldMatchValueChange: (index: number, severity: Severity, newMatchValue: string) => void;
+  isMappingChecked: boolean;
+  dataTestSubj?: string;
+  idAria?: string;
```
Thanks. Fixed.

nikitaindik commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/components/severity_mapping/severity_override.tsx:
```diff
@@ -0,0 +1,201 @@
+        <EuiFlexGroup
+          alignItems="center"
+          gutterSize="s"
+          onClick={!isDisabled ? onSeverityMappingChecked : noop}
```
Thanks. Fixed.

nikitaindik commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/components/severity_mapping/severity_override.tsx:
```diff
@@ -0,0 +1,201 @@
+        >
+          <EuiFlexItem grow={false}>
+            <EuiCheckbox
+              id={`severity-mapping-override`}
```
Thanks. Fixed.

nikitaindik commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/components/severity_mapping/severity_override.tsx:
```diff
@@ -0,0 +1,201 @@
+      }
+      fullWidth
+      data-test-subj={`${dataTestSubj}-severityOverride`}
+      describedByIds={idAria ? [idAria] : undefined}
```
Thanks. Fixed.

nikitaindik commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/components/severity_mapping/severity_override.tsx:
```diff
@@ -0,0 +1,201 @@
+      <NestedContent>
+        <EuiSpacer size="s" />
+        {isMappingChecked && (
+          <EuiFlexGroup direction={'column'} gutterSize="s">
```
Thanks. Fixed.

nikitaindik commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/components/severity_mapping/severity_override.tsx:
```diff
@@ -0,0 +1,201 @@
+                >
+                  <EuiFlexItemComboBoxColumn>
+                    <FieldComponent
+                      placeholder={''}
```
Thanks. Fixed.

nikitaindik commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/components/severity_mapping/severity_override.tsx:
```diff
@@ -0,0 +1,201 @@
+                    <FieldComponent
+                      placeholder={''}
+                      selectedField={getFieldTypeByMapping(severityMappingItem, indices)}
+                      isLoading={false}
```
Thanks. Fixed.

nikitaindik commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/components/severity_mapping/severity_override.tsx:
```diff
@@ -0,0 +1,201 @@
+                      isClearable={false}
+                      indexPattern={indices}
+                      onChange={onFieldChange.bind(null, index, severityMappingItem.severity)}
+                      data-test-subj={`detectionEngineStepAboutRuleSeverityMappingField-${severityMappingItem.severity}-${index}`}
```
Thanks. Fixed.

nikitaindik commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/components/severity_mapping/severity_override.tsx:
```diff
@@ -0,0 +1,201 @@
+                  <EuiFlexItemComboBoxColumn>
+                    <AutocompleteFieldMatchComponent
+                      autocompleteService={services.unifiedSearch.autocomplete}
+                      placeholder={''}
```
Thanks. Fixed.

nikitaindik commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/components/severity_mapping/severity_override.tsx:
```diff
@@ -0,0 +1,201 @@
+                      selectedValue={severityMappingItem.value}
+                      isClearable={true}
+                      isDisabled={isDisabled}
+                      isLoading={false}
```
Thanks. Fixed.

nikitaindik commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/components/severity_mapping/severity_override.tsx:
```diff
@@ -0,0 +1,201 @@
+                      isDisabled={isDisabled}
+                      isLoading={false}
+                      indexPattern={indices}
+                      onChange={onFieldMatchValueChange.bind(
```
Refactored this bit

nikitaindik commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/components/severity_mapping/severity_override.tsx:
```diff
@@ -0,0 +1,201 @@
+                        index,
+                        severityMappingItem.severity
+                      )}
+                      data-test-subj={`detectionEngineStepAboutRuleSeverityMappingValue-${severityMappingItem.severity}-${index}`}
```
Thanks. Fixed.

nikitaindik commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/components/severity_mapping/severity_override.tsx:
```diff
@@ -0,0 +1,201 @@
+                    />
+                  </EuiFlexItemComboBoxColumn>
+                  <EuiFlexItemIconColumn grow={false}>
+                    <EuiIcon type={'sortRight'} />
```
Thanks. Fixed.

nikitaindik commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/components/step_about_rule/schema.tsx:
```diff
@@ -129,6 +130,26 @@ export const schema: FormSchema<AboutStepRule> = {
+          ...args: Parameters<ValidationFunc>
+        ): ReturnType<ValidationFunc<{}, ERROR_CODE>> | undefined => {
+          const [{ value, path }] = args;
+          const fieldValue = value as { value: unknown };
```
Thanks. Fixed.

nikitaindik commented on x-pack/plugins/security_solution/public/detection_engine/rule_management/components/rule_details/three_way_diff/final_edit/fields/investigation_fields.tsx:
```diff
@@ -0,0 +1,82 @@
+}: InvestigationFieldsEditProps): JSX.Element {
+  const { type } = finalDiffableRule;
+
+  const defaultIndexPattern = useDefaultIndexPattern();
+  const indexPatternParameters = getUseRuleIndexPatternParameters(
+    finalDiffableRule,
+    defaultIndexPattern
+  );
+  const { indexPattern, isIndexPatternLoading } = useRuleIndexPattern(indexPatternParameters);
```
I haven't added it yet since your PR is still under review. I've included it to the list for later refactoring.

nikitaindik commented on x-pack/plugins/security_solution/public/detection_engine/rule_management/components/rule_details/three_way_diff/final_edit/fields/risk_score_mapping.tsx:
```diff
@@ -0,0 +1,106 @@
+  const { value, setValue } = field;
+
+  const handleToggleMappingChecked = useCallback(() => {
+    setValue({
```
Refactored here and in other places

nikitaindik commented on x-pack/plugins/security_solution/public/detection_engine/rule_management/components/rule_details/three_way_diff/final_edit/fields/rule_name_override.tsx:
```diff
@@ -0,0 +1,71 @@
+      fieldType: 'string',
+      indices: indexPattern,
+      isDisabled: isIndexPatternLoading,
+      placeholder: '',
```
Fixed

nikitaindik commented:
@maximpn Thanks for the thorough testing!

> Max Signals allows to save any value despite a validation error show

I fixed this one.

> Rule name override doesn't show any values (there is no data in rule indices locally so Kibana isn't able to fetch available values)

> Timestamp override doesn't show any values (there is no data in rule indices locally so Kibana isn't able to fetch available values)

These two issues are not easy to fix due to the way the dropdown component is built. Let's discuss how we should approach this offline.

elasticmachine commented:
Pinging @elastic/security-detections-response (Team:Detections and Resp)

elasticmachine commented:
Pinging @elastic/security-solution (Team: SecuritySolution)

elasticmachine commented:
Pinging @elastic/security-detection-rule-management (Team:Detection Rule Management)

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/249592)
* Commit: b4a4c29a4b0ce24e85fb9abe0d2bdbc31abe3b26

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/9f514f641d8e41c8807781eaa407c5a9cbca4f95) | [after](https://github.com/elastic/kibana/commit/b4a4c29a4b0ce24e85fb9abe0d2bdbc31abe3b26) | diff |
| --- | --- | --- | --- |
| `securitySolution` | 6131 | 6159 | +28 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/9f514f641d8e41c8807781eaa407c5a9cbca4f95) | [after](https://github.com/elastic/kibana/commit/b4a4c29a4b0ce24e85fb9abe0d2bdbc31abe3b26) | diff |
| --- | --- | --- | --- |
| `lists` | 143.8KB | 143.8KB | +3.0B |
| `securitySolution` | 13.4MB | 13.4MB | +13.7KB |
| total |  |  | +13.7KB |

<details><summary>Unknown metric groups</summary>

#### ESLint disabled line counts
| id | [before](https://github.com/elastic/kibana/commit/9f514f641d8e41c8807781eaa407c5a9cbca4f95) | [after](https://github.com/elastic/kibana/commit/b4a4c29a4b0ce24e85fb9abe0d2bdbc31abe3b26) | diff |
| --- | --- | --- | --- |
| `securitySolution` | 553 | 554 | +1 |

#### Total ESLint disabled count
| id | [before](https://github.com/elastic/kibana/commit/9f514f641d8e41c8807781eaa407c5a9cbca4f95) | [after](https://github.com/elastic/kibana/commit/b4a4c29a4b0ce24e85fb9abe0d2bdbc31abe3b26) | diff |
| --- | --- | --- | --- |
| `securitySolution` | 639 | 640 | +1 |


</details>


### History
* :broken_heart: [Build #249536](https://buildkite.com/elastic/kibana-pull-request/builds/249536) failed 6986e21bd8362b1b15dec65018798ff39f29e7c6
* :broken_heart: [Build #249511](https://buildkite.com/elastic/kibana-pull-request/builds/249511) failed db15ed8312834de4ef8c504e6cbef7288d859bb6
* :broken_heart: [Build #247775](https://buildkite.com/elastic/kibana-pull-request/builds/247775) failed fcd9457b531dd1214b11935474c21ebfa167031e

cc @nikitaindik

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/249592","number":249592,"commit":"b4a4c29a4b0ce24e85fb9abe0d2bdbc31abe3b26"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/249536","number":249536,"commit":"6986e21bd8362b1b15dec65018798ff39f29e7c6"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/249511","number":249511,"commit":"db15ed8312834de4ef8c504e6cbef7288d859bb6"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/247775","number":247775,"commit":"fcd9457b531dd1214b11935474c21ebfa167031e"}],"number":249592}
buildkite-pr-comment-->

nikitaindik committed:
Make `AutocompleteField` display values that are missing from indices

nikitaindik committed:
Merge remote-tracking branch 'upstream/main' into editable-common-fields-2

nikitaindik commented:
@maximpn UPD
> These two issues are not easy to fix due to the way the dropdown component is built. Let's discuss how we should approach this offline.

I've used the approach that we discussed offline and fixed these two as well in https://github.com/elastic/kibana/pull/196642/commits/66afe182e36364f7c51f188b4223ba64e37c90e1

maximpn commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/components/autocomplete_field/index.tsx:
```diff
@@ -38,15 +40,19 @@ export const AutocompleteField = ({
+  const { selectedField, indexPattern } = useMemo(() => {
+    /* eslint-disable @typescript-eslint/no-shadow */
+    /* If the field is not found in the indices, we still want to show it in the dropdown */
+    const selectedField = foundField || { name: field.value, type: fieldType };
```
It's better to use a different name inside `useMemo` and remove `eslint-disable`.

maximpn commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/components/autocomplete_field/index.tsx:
```diff
@@ -30,6 +30,8 @@ export const AutocompleteField = ({
   fieldType,
   placeholder,
 }: AutocompleteFieldProps) => {
+  const fieldTypeFilter = useMemo(() => [fieldType], [fieldType]);
```
`fieldType` has `string` type. It doesn't look beneficial to memoize it.

maximpn commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/components/autocomplete_field/index.tsx:
```diff
@@ -38,15 +40,19 @@ export const AutocompleteField = ({
-    );
-    return newSelectedField;
-  }, [field.value, indices]);
+  const foundField = field.value && indices.fields.find(({ name }) => name === field.value);
```
This line should be inside `useMemo()`'s callback.

maximpn commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/components/autocomplete_field/index.tsx:
```diff
@@ -38,15 +40,19 @@ export const AutocompleteField = ({
+  const foundField = field.value && indices.fields.find(({ name }) => name === field.value);
 
-  const fieldTypeFilter = useMemo(() => [fieldType], [fieldType]);
+  const { selectedField, indexPattern } = useMemo(() => {
+    /* eslint-disable @typescript-eslint/no-shadow */
+    /* If the field is not found in the indices, we still want to show it in the dropdown */
+    const selectedField = foundField || { name: field.value, type: fieldType };
+    const indexPattern = foundField
+      ? indices
+      : { ...indices, fields: [...indices.fields, selectedField] };
+
+    return { selectedField, indexPattern };
+    /* eslint-enable @typescript-eslint/no-shadow */
+  }, [field.value, indices, foundField, fieldType]);
```
For better readability this logic could be moved to an "internal" hook named like `useSelectedField()` and placed under the component in the file.

Though it might be better to have this logic in `FieldComponent`.

maximpn reviewed (commented):
@nikitaindik I had a look at the recent changes and left comments. I think we should make slight improvements before merging it back.

nikitaindik committed:
Simplify logic in `AutocompleteField` and rename `AutocompleteField` to `EsFieldSelectorField`

nikitaindik committed:
Merge remote-tracking branch 'upstream/main' into editable-common-fields-2

nikitaindik commented:
@maximpn Thanks for taking another look Maxim and discussing it with me over Zoom. I've pushed a [commit](https://github.com/elastic/kibana/pull/196642/commits/e276eb697ba0018b08f46bd8fb9d2c801b448d05) that addresses your feedback and also:
 - Renames `AutocompleteField` to `EsFieldSelectorField`
 - Renames `FieldComponent` to `EsFieldSelector`
 - Renames `useField` to `useEsField`

Please take a look.

nkhristinin reviewed (commented):
I tested locally, updated flow and create/edit rule. LGTM

Left some question about implementations

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/249931)
* Commit: 892e1eb23fe9a910cfaa4fa0d5dff59924fe8334

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/99160f52d6a2e0ad4698ca8e5ac40316ebb733c9) | [after](https://github.com/elastic/kibana/commit/892e1eb23fe9a910cfaa4fa0d5dff59924fe8334) | diff |
| --- | --- | --- | --- |
| `securitySolution` | 6135 | 6163 | +28 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/99160f52d6a2e0ad4698ca8e5ac40316ebb733c9) | [after](https://github.com/elastic/kibana/commit/892e1eb23fe9a910cfaa4fa0d5dff59924fe8334) | diff |
| --- | --- | --- | --- |
| `lists` | 144.9KB | 144.9KB | -15.0B |
| `securitySolution` | 13.4MB | 13.4MB | +13.7KB |
| total |  |  | +13.6KB |

#### Public APIs missing exports
> Total count of every type that is part of your API that should be exported but is not. This will cause broken links in the API documentation system. Target amount is 0. `Run node scripts/build_api_docs --plugin [yourplugin] --stats exports` for more detailed information.


| id | [before](https://github.com/elastic/kibana/commit/99160f52d6a2e0ad4698ca8e5ac40316ebb733c9) | [after](https://github.com/elastic/kibana/commit/892e1eb23fe9a910cfaa4fa0d5dff59924fe8334) | diff |
| --- | --- | --- | --- |
| `@kbn/securitysolution-autocomplete` | 1 | 0 | -1 |

<details><summary>Unknown metric groups</summary>

#### ESLint disabled line counts
| id | [before](https://github.com/elastic/kibana/commit/99160f52d6a2e0ad4698ca8e5ac40316ebb733c9) | [after](https://github.com/elastic/kibana/commit/892e1eb23fe9a910cfaa4fa0d5dff59924fe8334) | diff |
| --- | --- | --- | --- |
| `securitySolution` | 553 | 554 | +1 |

#### Total ESLint disabled count
| id | [before](https://github.com/elastic/kibana/commit/99160f52d6a2e0ad4698ca8e5ac40316ebb733c9) | [after](https://github.com/elastic/kibana/commit/892e1eb23fe9a910cfaa4fa0d5dff59924fe8334) | diff |
| --- | --- | --- | --- |
| `securitySolution` | 639 | 640 | +1 |


</details>


### History
* :green_heart: [Build #249592](https://buildkite.com/elastic/kibana-pull-request/builds/249592) succeeded b4a4c29a4b0ce24e85fb9abe0d2bdbc31abe3b26
* :broken_heart: [Build #249536](https://buildkite.com/elastic/kibana-pull-request/builds/249536) failed 6986e21bd8362b1b15dec65018798ff39f29e7c6
* :broken_heart: [Build #249511](https://buildkite.com/elastic/kibana-pull-request/builds/249511) failed db15ed8312834de4ef8c504e6cbef7288d859bb6
* :broken_heart: [Build #247775](https://buildkite.com/elastic/kibana-pull-request/builds/247775) failed fcd9457b531dd1214b11935474c21ebfa167031e

cc @nikitaindik

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/249931","number":249931,"commit":"892e1eb23fe9a910cfaa4fa0d5dff59924fe8334"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/249592","number":249592,"commit":"b4a4c29a4b0ce24e85fb9abe0d2bdbc31abe3b26"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/249536","number":249536,"commit":"6986e21bd8362b1b15dec65018798ff39f29e7c6"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/249511","number":249511,"commit":"db15ed8312834de4ef8c504e6cbef7288d859bb6"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/247775","number":247775,"commit":"fcd9457b531dd1214b11935474c21ebfa167031e"}],"number":249931}
buildkite-pr-comment-->

maximpn reviewed (approved):
@nikitaindik Thanks for addressing my comments 🙏 

I tested the PR on the latest commit and editable components work as expected 👍

maximpn commented:
@nkhristinin There is time constrain on Prebuilt Rule Customization readiness. Do you think there is something critical in you comments and should be addressed in this PR? I created a ticket for rule creation/editing forms improvements. Let's add any desired enhancements/improvements to that ticket.

nkhristinin commented:
@maximpn  I think we can add tests for this function later https://github.com/elastic/kibana/pull/196642#discussion_r1836559759

Other comments, not blocking, more requesting reason for those changes

kibanamachine commented:
Starting backport for target branches: 8.x

https://github.com/elastic/kibana/actions/runs/11794964800

kibanamachine commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.x|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/199743">](https://github.com/elastic/kibana/pull/199743)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 9.4.3 -->

nikitaindik commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/components/autocomplete_field/index.tsx:
```diff
@@ -38,15 +40,19 @@ export const AutocompleteField = ({
-  const fieldTypeFilter = useMemo(() => [fieldType], [fieldType]);
+  const { selectedField, indexPattern } = useMemo(() => {
+    /* eslint-disable @typescript-eslint/no-shadow */
+    /* If the field is not found in the indices, we still want to show it in the dropdown */
```
Actually, I ended up removing this part and reverting to previous behaviour. It turned out we already handle such case in a nested component.

nikitaindik commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation/components/required_fields/required_fields_row.tsx:
```diff
@@ -44,14 +44,6 @@ export const RequiredFieldRow = ({
   const rowFieldConfig: FieldConfig<RequiredField | RequiredFieldInput, {}, RequiredFieldInput> =
     useMemo(
       () => ({
-        deserializer: (value) => {
```
It was needed to convert the type from `RequiredField` to `RequiredFieldInput`, but I've modified the component to use `RequiredField`, so this conversion is no longer needed.

nikitaindik commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation/components/required_fields/make_validate_required_field.ts:
```diff
@@ -51,3 +51,20 @@ export function makeValidateRequiredField(parentFieldPath: string) {
+) {
+  /*
+    Getting values for required fields via flattened fields instead of using `getFormData`.
+    This is because `getFormData` applies a serializer function to field values, which might update values.
```
[Serializer](https://docs.elastic.dev/form-lib/examples/serializers-deserializers) is a function you can pass to the form library. It converts form data from an internal format to an output format when you click Submit. 

To get the form data for validation, we normally use `form.getFormData()`. However, this method internally calls the serializer. In this case, we need the form data in its internal format, before it is serialized, because the returned value would be different if you pass a different serialzers.
